skimage/segmentation/_join.py

```diff
@@ -116,26 +117,179 @@ def relabel_sequential(label_field, offset=1):
+    offset = int(offset)
+    in_vals = np.unique(label_field)
+    if in_vals[0] == 0:
+        # always map 0 to 0
+        out_vals = np.concatenate(
+            [[0], np.arange(offset, offset+len(in_vals)-1)]
+        )
+    else:
+        out_vals = np.arange(offset, offset+len(in_vals))
+    input_type = label_field.dtype
+
+    # Some logic to determine the output type:
+    #  - we don't want to return a smaller output type than the input type,
+    #  ie if we get uint32 as labels input, don't return a uint8 array.
+    #  - but, in some cases, using the input type could result in overflow. The
+    #  input type could be a signed integer (e.g. int32) but
+    #  `np.min_scalar_type` will always return an unsigned type. We check for
+    #  that by casting the largest output value to the input type. If it is
+    #  unchanged, we use the input type, else we use the unsigned minimum
+    #  required type
+    required_type = np.min_scalar_type(out_vals[-1])
+    if input_type.itemsize < required_type.itemsize:
         output_type = required_type
-    forward_map = np.zeros(max_label + 1, dtype=output_type)
-    forward_map[labels0] = new_labels0
-    inverse_map = np.zeros(new_max_label + 1, dtype=output_type)
-    inverse_map[offset:] = labels0
-    relabeled = forward_map[label_field]
-    return relabeled, forward_map, inverse_map
+    else:
+        if input_type.type(out_vals[-1]) == out_vals[-1]:
+            output_type = input_type
+        else:
+            output_type = required_type
+    out_array = np.empty(label_field.shape, dtype=output_type)
+    out_vals = out_vals.astype(output_type)
+    map_array(label_field, in_vals, out_vals, out=out_array)
+    fw_map = ArrayMap(in_vals, out_vals)
+    inv_map = ArrayMap(out_vals, in_vals)
+    return out_array, fw_map, inv_map
+
+
+def map_array(input_arr, input_vals, output_vals, out=None):
+    """Map values from input array from input_vals to output_vals.
+
+    Parameters
+    ----------
+    input_arr : array of int, shape (M[, N][, P][, ...])
+        The input label image.
+    input_vals : array of int, shape (N,)
+        The values to map from.
+    output_vals : array, shape (N,)
+        The values to map to.
+    out: array, same shape as `input_arr`
+        The output array. Will be created if not provided. It should
+        have the same dtype as `output_vals`.
+
+    Returns
+    -------
+    out : array, same shape as `input_arr`
+        The array of mapped values.
+    """
+
+    # We ravel the input array for simplicity of iteration in Cython:
+    orig_shape = input_arr.shape
+    # NumPy docs for `np.ravel()` says:
+    # "When a view is desired in as many cases as possible, 
+    # arr.reshape(-1) may be preferable."
+    input_arr = input_arr.reshape(-1)
+    if out is None:
+        out = np.empty_like(input_arr, dtype=output_vals.dtype)
+    elif not (out.flags['FORC'] or out.ndim == 1):
```

Comment:
I wonder if the check for `out` to be 1D is relevant since it is reshaped later :-/

Reply:
If it is already 1d we don't care if it is contiguous. I think Cython takes care of iterating over non-contiguous arrays correctly. 

TBH, I prefer the original code: reshape first, complain if we got a copy (rather than a view) afterwards. It makes the intention clearer. The point is simply that we need a 1D - view into the source array to modify it in place with the Cython code. If numpy returns a copy instead (for whatever reason, maybe this could happen even if `out.flags['FORC']` is `True` if some other conditions are satisfied) we would not catch it with the logic that is in place now.

Reply:
@VolkerH you are right. Indeed I had to look up what `out.flags['FORC']` meant. Having said that, the earlier test was not exactly comprehensive either and would not have caught a copy if the array was contiguous to begin with. Perhaps we should use `np.may_share_memory`.

Reply:
@VolkerH, I proposed the modifications above simply to reflect the raised error message, nothing more... If the test isn't appropriate, it may be means that the error message is not correct!
Lookin at note section in the [`np.reshape`](https://numpy.org/doc/stable/reference/generated/numpy.reshape.html?highlight=reshape#numpy-reshape) documentation, one solution may be:
```python
try:
    out_view = out.view()
    out_view.shape = (-1, )  # this will fail if a copy is necessary
except AttributeError:
    raise ValueError("An appropriate error message")
```

Reply:
Whoa. Nice idea, @rfezzani!